### PR TITLE
perf(traverse): reduce string operations in `get_var_name_from_node`

### PR DESCRIPTION
**Pull request by overlookmotel**
Improve the performance of creating a var name from an AST node.

## Perf

2 different optimizations:

#### 1. Avoid building a huge string

Previously we concatenated all identifiers into a string of potentially large length, then truncated it to 20 bytes at the end.

Instead, once the string reaches the maximum length, stop adding to it. This avoids pointless work, and avoids extra allocations to grow the `String`.

#### 2. Remove "needs to grow?" checks

The first optimization ensures that the `String` never exceeds 20 bytes. Utilize this invariant to remove "does the `String` need to grow?" checks on every push to the `String`.

This uses the `push_unchecked` and `push_str_unchecked` methods added to `String` in #24006.

Adverserial review by Claude found the implementation to be sound, and it performed fuzzing using 500,000 random cases to test the implementation. Tests which are added in this PR also cover all the edge cases.

## Fix

This PR also fixes incorrect truncation when some of the parts contain Unicode characters.

Previously the string would be truncated to be shorter than 20 bytes when the last char before the cut point is a complete multi-byte (Unicode) character. Now it checks whether to truncate to less than 20 bytes based on whether the 21st byte is a character sequence start byte, instead of whether the 20th byte is a continuation character.

The tests are altered to reflect this fix.

### CODE REVIEW

**Comment by overlookmotel** [comment]
* **#24007** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/24007?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/oxc-project/oxc/24007?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#24006** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/24006?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent changes, fast-track this PR to the front of the merge queue

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=24007).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: crates/oxc_traverse/src/ast_operations/gather_node_parts.rs [hunk 1]
```diff
@@ -5,43 +5,117 @@
...
+            let bytes = part.as_bytes();
+
+            // SAFETY:
+            // * `name.len() + part.len() > MAX_LEN` (`new_len > MAX_LEN` check above)
+            // * `name.len() <= MAX_LEN` (`name.len() < MAX_LEN` at top of closure, and we pushed at most 1 `$`)
+            // * Therefore: `part.len() > MAX_LEN - name.len()`
+            // * `cut_at = MAX_LEN - name.len()`
+            // * Therefore: `part.len() > cut_at`
+            // * Reversed: `cut_at < part.len()`
+            // * `cut_at` is in bounds of `bytes`
```
[the remark this answers is not in the document]
Comment (overlookmotel) [reply crates/oxc_traverse/src/ast_operations/gather_node_parts.rs:68]: Untrue. Slices cannot have length exceeding `isize::MAX`, so adding the lengths of 2 slices cannot overflow `usize`.

I've added a comment.

**Review by Dunqing — APPROVED** [review]

**Comment by Dunqing** [comment]
### Merge activity

* **Jul 2, 1:28 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Jul 2, 1:28 AM UTC**: `Dunqing` added this pull request to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc).
* **Jul 2, 1:52 AM UTC**: Merged by the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc).
